File: Medium/215_KthLargestElementinAnArray.js
/**
 * pseudo-code:
 * function QuickSelect(nums[1..n], k) // returns element of rank k in nums, 1 <= k <= n
 * if n > 1 then
 *   Choose pivot element p = nums[Math.floor(nums.length / 2)] // mid value
 *   Let Lesser be an array of all elements from nums less than p
 *   Let Greater be an array of all elements from nums greater than p
 *   Let equal be the occurrence times of p
 *   if |Greater| = k - 1 or (|Greater| < k - 1 and |Greater| + equal > k - 1)
 *      return p
 *   Else if |Greater| > k - 1
 *      return QuickSelect(Greater, k)
 *   Else if |Greater| < k - 1
 *      return QuickSelect(Lesser, k - |Greater| - equal)
 * else 
 *   return nums[1]
 */


/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const findKthLargest = function (nums, k) {
    if (nums.length > 1) {
        const pivot = nums[Math.floor(nums.length / 2)];
        let lesser = [];
        let greater = [];
        let equal = 0;
        for (const num of nums) {
            if (num < pivot) {
                lesser.push(num);
            } else if (num > pivot) {
                greater.push(num);
            } else {
                equal++;
            }
        }

        delete nums;

        if (greater.length === k - 1 ||
            (greater.length < k - 1 && greater.length + equal > k - 1)) {
            return pivot;
        } else if (greater.length > k - 1) {
            lesser = [];
            return findKthLargest(greater, k);
        } else if (greater.length < k - 1) {
            const greaterLength = greater.length;
            delete greater;
            return findKthLargest(lesser, k - greaterLength - equal);
        }
    } else {
        return nums[0];
    }
};

File: Medium/215_kth-largest-element-in-an-array.java
// With heap sort 
class Solution {
    public int findKthLargest(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return -1;
        }

        int heapSize = nums.length;

        // build max heap
        buildMaxHeap(nums);

        // sort the k - 1 largest elements
        for (int i = nums.length - 1; i >= nums.length - k + 1; i--) {
            swap(nums, 0, i);
            heapSize--;
            maxHeapify(nums, 0 , heapSize);
        }
        
        return nums[0];
    }

    private void buildMaxHeap(int[] nums) {
        for (int i = nums.length / 2; i >= 0; i--) {
            maxHeapify(nums, i, nums.length);
        }
    }

    private void maxHeapify(int[] nums, int i, int heapSize) {
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        int largest = i;
        if (l < heapSize && nums[l] > nums[largest]) {
            largest = l;
        } else {
            largest = i;
        }

        if (r < heapSize && nums[r] > nums[largest]) {
            largest = r;
        }

        if (largest != i) {
            swap(nums, i , largest);
            maxHeapify(nums, largest, heapSize);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

// With heap java api approach
class Solution {
    public int findKthLargest(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return -1;
        }

        // max heap
        Queue<Integer> heap = new PriorityQueue<>((a, b) -> b - a);

        for (int num : nums) {
            heap.offer(num);
        }

        // delete the peak elements k - 1 times
        for (int i = 0; i < k - 1; i++) {
            heap.poll();
        }

        // obtain the kth largest element
        return heap.peek();
    }
}


class Solution {
    public int findKthLargest(int[] nums, int k) {
        int pivot = nums.length - k + 1;
        return quickSelect(0, nums.length - 1, pivot, nums);
    }

    private int quickSelect(int start, int end, int k, int[] nums) {
        // base case
        if (start == end) {
            return nums[start];
        }
        int index = partition(nums, start, end);

        // including nums[index]
        int num = index - start + 1;
        if (k == num) {
            return nums[index];
        } else if (k < num) {
            return quickSelect(start, index - 1, k, nums);
        } else {
            return quickSelect(index + 1, end, k - num, nums);
        }
    }

    private int partition(int[] nums, int start, int end) {
        int pivot = start; // pivot = end;
        int value = nums[pivot];

        while (start < end) {
            while (start < end && nums[end] >= value) {
                end--;
            }

            // Find the element that smaller than value;
            // nums[start] = nums[end];
            swap(nums, start, end);

            while (start < end && nums[start] <= value) {
                start++;
            }

            // Find the element that smaller than value;
            // nums[end] = nums[start];
            swap(nums, end, start);
        }
        // nums[start] = value;
        return start;
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}